src/uct/ib/base/ib_md.h

```diff
@@ -558,26 +562,56 @@ static UCS_F_ALWAYS_INLINE uint32_t uct_ib_memh_get_lkey(uct_mem_h memh)
 }
 
 
+static UCS_F_ALWAYS_INLINE UCS_F_MAYBE_UNUSED ucs_status_t
+uct_ib_md_rkey_mem_dereg_invalidate_check(uint32_t rkey, uint32_t access_mask,
+                                          uint64_t cap_mask)
+{
+    if (access_mask) {
+        if (!cap_mask) {
+            return UCS_ERR_UNSUPPORTED;
+        } else if (rkey == UCT_IB_INVALID_MKEY) {
+            return UCS_ERR_INVALID_PARAM;
+        }
+    }
+
+    return UCS_OK;
+}
+
+
 static UCS_F_ALWAYS_INLINE ucs_status_t
 uct_ib_md_mem_dereg_params_invalidate_check(
-        const uct_md_mem_dereg_params_t *params)
+        const uct_md_mem_dereg_params_t *params, uint64_t cap_flags)
 {
+#if ENABLE_PARAMS_CHECK
     uct_ib_mem_t *ib_memh;
     unsigned flags;
+    ucs_status_t status;
 
-    if (ENABLE_PARAMS_CHECK) {
-        ib_memh = (uct_ib_mem_t*)UCT_MD_MEM_DEREG_FIELD_VALUE(params, memh,
-                                                              FIELD_MEMH, NULL);
-        flags   = UCT_MD_MEM_DEREG_FIELD_VALUE(params, flags, FIELD_FLAGS, 0);
-        if ((flags & UCT_MD_MEM_DEREG_FLAG_INVALIDATE) &&
-            (ib_memh->indirect_rkey == UCT_IB_INVALID_MKEY)) {
-            return UCS_ERR_INVALID_PARAM;
-        }
+    ib_memh = (uct_ib_mem_t*)UCT_MD_MEM_DEREG_FIELD_VALUE(params, memh,
+                                                          FIELD_MEMH, NULL);
+    flags   = UCT_MD_MEM_DEREG_FIELD_VALUE(params, flags, FIELD_FLAGS, 0);
+    if (!(flags & UCT_MD_MEM_DEREG_FLAG_INVALIDATE)) {
+        return UCS_OK;
     }
 
+    status = uct_ib_md_rkey_mem_dereg_invalidate_check(
+            ib_memh->indirect_rkey,
+            ib_memh->flags & UCT_IB_MEM_ACCESS_REMOTE_RMA,
+            cap_flags & UCT_MD_FLAG_INVALIDATE_RMA);
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    return uct_ib_md_rkey_mem_dereg_invalidate_check(
+            ib_memh->atomic_rkey,
+            ib_memh->flags & UCT_IB_MEM_ACCESS_REMOTE_ATOMIC,
+            cap_flags & UCT_MD_FLAG_INVALIDATE_AMO);
```

Comment:
these checks are redundant IMO
we want that even when only RMA invalidate is supported, we are still able to register uct_memh with both AMO|RMA and invalidate. in this case we expect remote RMA will really be invalidated, but remote AMO, if have such, could have undefined behavior.

Reply:
IMO it's API restriction that we don't need to add since it will make UCP code more complex